### ISSUE TITLE
[request]: Enforce Content-Length on pre-signed PutObject (S3)

### ISSUE THREAD

**Issue by stefanhoelzl** [awslabs/aws-sdk-rust#466]
### Community Note
* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue, please leave a comment

### Tell us about your request

I want to be able to enforce a Content-Length for a pre-signed PutObject URL, like it is possible with the Go and Ruby SDK (see https://github.com/aws/aws-sdk-go/issues/625).

When generating the pre-signed URL I want to be able to specify a content-length.
When using the pre-signed URL this content-length should be enforced, so that it is not possible to upload files where the file size does not match the specified content-length.

### Tell us about the problem you're trying to solve.

I want to make sure that my users who get pre-signed URL from my service can only upload files with a certain file size.

### Are you currently working around this issue?

For testing purposes I do not enforce the content-length, but that is not a solution for production.

### Additional context

With some guidance I am willing to help implementing this feature.

**Comment by stefanhoelzl** [comment]
I was digging a bit into [smithy-rs](https://github.com/awslabs/smithy-rs/) and found [this piece of code](https://github.com/awslabs/smithy-rs/blob/main/aws/rust-runtime/aws-inlineable/src/presigning.rs#L268)
```rust
// Remove headers from input serialization that shouldn't be part of the presigned
// request since the request body is unsigned and left up to the person making the final
// HTTP request.
req.headers_mut().remove(CONTENT_LENGTH);
req.headers_mut().remove(CONTENT_TYPE);
```
Am I right, that this is the part which prevents me from enforcing a content-length on my signed urls?

**Comment by stefanhoelzl** [comment]
tested it locally, and it was not the right spot to change.

But I found [the right one](https://github.com/awslabs/smithy-rs/blob/main/aws/rust-runtime/aws-sigv4/src/http_request/canonical_request.rs#L228
)
```rust
// Exclude content-length and content-type for query param signatures since the
// body is unsigned for these use-cases, and the size is not known up-front.
if name == CONTENT_LENGTH || name == CONTENT_TYPE {
    continue;
}
```

My proposal would be that `CONTENT_LENGTH` is only excluded if its value is equal to `0`.

This would mean that there is not change in behavior when no content length is set during creation of the request
```rust
client
    .put_object()
    .bucket(bucket)
    .key(key)
    .presigned(PresigningConfig::expires_in(core::time::Duration::from_secs(60)).unwrap())
```

but it gets included in the signed headers, when a content-length was explicitly given
```rust
client
    .put_object()
    .bucket(bucket)
    .key(key)
    .content_length(4)
    .presigned(PresigningConfig::expires_in(core::time::Duration::from_secs(60)).unwrap())
```

I am happy to open a PR for this changes!

**Comment by Velfi** [comment]
This might actually requires some special handling. Our team is going to do some more research on the best way to support this so hold tight.

**Comment by jdisanti** [comment]
Thanks for reporting!

This is a bug in the presigning implementation and will require a bit of refactoring of the code generator and query param SigV4 signing implementation to fix. I'm working on it.

**Comment by jdisanti** [comment]
The fix for this went out in [v0.8.0](https://github.com/awslabs/aws-sdk-rust/releases/tag/v0.8.0).

**Comment by stefanhoelzl** [comment]
Thank you @jdisanti and @Velfi for fixing and releasing this so quickly.

I just verified that version 0.8.0 is working as expected in my application.
